add an explicit Broker creation not relying on defaults (#2756)
knative/docs: docs/eventing/broker/README.md

EOF
```

> ## Explicit Broker configuration

Reviewer: ```suggestion
## Configuring broker classes
```

> You can of course create a Broker without relying on the defaults (as described
> above) by specifying the `BrokerClass` by using an annotation to specify which
> Broker implementation you'd like to use:

Reviewer: ```suggestion
You can configure Knative Eventing so that when you create a broker, it uses a different type of broker than the default Knative channel-based broker.

To configure a different broker type, or *class*, you must modify the `eventing.knative.dev/broker.class` annotation and `spec.config` for the Broker object.

`MTChannelBasedBroker` is the broker class default.
```

Author: kafka, gcp, hopefully soon rabbit too.

Reviewer: Can you add maybe an example here that doesn't use MT if that's the default, i.e. change the example to Kafka or something? Can we also list all the available options?

Q: Are there examples of different ConfigMaps for all of these, can we change the `spec.config` in the example too so that it points to a different ConfigMap, or is that default in the sense that you have to configure it first to define the default channel type, rather than default meaning InMemoryChannel or something?

> ```yaml
> kind: Broker
> metadata:
>   annotations:
>     eventing.knative.dev/broker.class: MTChannelBasedBroker
> ```

Reviewer: ```suggestion
### Procedure

1. Modify the `eventing.knative.dev/broker.class` annotation. Replace `MTChannelBasedBroker` with the class type you want to use:

```yaml
kind: Broker
metadata:
  annotations:
    eventing.knative.dev/broker.class: MTChannelBasedBroker
```
```

> and explicitly configuring the spec.config, for example:
> ```yaml
> kind: Broker
> spec:
>   config:
>     apiVersion: v1
>     kind: ConfigMap
>     name: config-br-default-channel
>     namespace: knative-eventing
> ```

Reviewer: ```suggestion
1. Configure the `spec.config` with the details of the ConfigMap that defines the backing channel for the broker class:

```yaml
kind: Broker
spec:
  config:
    apiVersion: v1
    kind: ConfigMap
    name: config-br-default-channel
    namespace: knative-eventing
```
```

A full example could look something like this:

```shell
kubectl create -f - <<EOF
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  annotations:
    eventing.knative.dev/broker.class: MTChannelBasedBroker
  name: default
  namespace: default
spec:
  config:
    apiVersion: v1
    kind: ConfigMap
    name: config-br-default-channel
    namespace: knative-eventing
EOF

> ```

Reviewer: Move this to the intro part maybe?

Author: Ok, so we have the two steps you can modify, class / config. Then showing them combined? How's that?

Reviewer: I think it's actually fine as is now I see it with the other changes :slightly_smiling_face: thank you!

Reviewer: One last thing - I would actually remove the `kubectl create -f - <<EOF` bits so that it's just showing the YAML example. That's up to you though :slightly_smiling_face:

Author: ok, done :)

EOF
```

## Configuring broker classes

You can configure Knative Eventing so that when you create a broker, it uses a different type of broker than the default Knative channel-based broker.

To configure a different broker type, or *class*, you must modify the `eventing.knative.dev/broker.class` annotation and `spec.config` for the Broker object.

`MTChannelBasedBroker` is the broker class default.

### Procedure

1. Modify the `eventing.knative.dev/broker.class` annotation. Replace `MTChannelBasedBroker` with the class type you want to use:

```yaml
kind: Broker
metadata:
  annotations:
    eventing.knative.dev/broker.class: MTChannelBasedBroker

1. Configure the `spec.config` with the details of the ConfigMap that defines the backing channel for the broker class:

```yaml
kind: Broker
spec:
  config:
    apiVersion: v1
    kind: ConfigMap
    name: config-br-default-channel
    namespace: knative-eventing

> 

Reviewer: you might also need to fix the indentation for the steps, idk how they'll render - maybe build a preview if you can? :slightly_smiling_face:

EOF
```

## Configuring broker classes

You can configure Knative Eventing so that when you create a broker, it uses a
different type of broker than the default Knative channel-based broker. To
configure a different broker type, or *class*, you must modify the
`eventing.knative.dev/broker.class` annotation and `spec.config` for the Broker
object. `MTChannelBasedBroker` is the broker class default.

### Procedure

1. Modify the `eventing.knative.dev/broker.class` annotation. Replace
`MTChannelBasedBroker` with the class type you want to use:

```yaml
kind: Broker
metadata:
  annotations:
    eventing.knative.dev/broker.class: MTChannelBasedBroker
```

1. Configure the `spec.config` with the details of the ConfigMap that defines

> the backing channel for the broker class: 

Reviewer 2: <!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
the backing channel for the broker class:
```
